### ISSUE TITLE
Is g_device_cuda_lock_arrays static?

### ISSUE THREAD

**Issue by hfinkel** [kokkos/kokkos#2235]
https://github.com/kokkos/kokkos/blob/83799feaae8c4ac32db83044fafd139732493911/core/src/Cuda/Kokkos_Cuda_Locks.hpp#L109

There's this comment:

```
/// When relocatable device code is disabled, an instance of this variable
/// will be created in every translation unit that sees this header file
/// (we make this clear by marking it static, meaning no other translation
///  unit can link to it).
```

But, this comment seems stale. The actual definition reads:

```
__device__
#ifdef KOKKOS_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE
    __constant__ extern
#endif
    Kokkos::Impl::CudaLockArrays g_device_cuda_lock_arrays;
```

which is explicitly extern, not static.

Then, later we have this:

```
/* Dan Ibanez: it is critical that this code be a macro, so that it will
   capture the right address for Kokkos::Impl::g_device_cuda_lock_arrays!
   putting this in an inline function will NOT do the right thing! */
#define KOKKOS_COPY_CUDA_LOCK_ARRAYS_TO_DEVICE()                      
```
but this comment does not explain why an inline function will not work. My guess is that, when `g_device_cuda_lock_arrays` was static, this could not be inline because of ODR (but that doesn't explain why it could not have been a static function), but it's not static currently. Thus, maybe this comment is stale too.

**Comment by crtrott** [comment]
So the definition as it relates to the comment which says "When relocatable device code is DISabled" is __device__ but not static. It used to be `static __device__ __constant__` then became `__device__ __constant__` and finally `__device__` due to issues with clang CUDA, newer NVCC versions and performance issues. Regarding the inline thing: a pure inline didn't work for some weird reason I don't remember. Essentially we had in the field errors we tracked down to this @ibaned may remember more about it. Most of the problems were anyway specific to NVCC and not Clang ...

**Comment by hfinkel** [comment]
> So the definition as it relates to the comment which says "When relocatable device code is DISabled" is **device** but not static. It used to be `static __device__ __constant__` then became `__device__ __constant__` and finally `__device__` due to issues with clang CUDA, newer NVCC versions and performance issues.

I agree. I was pointing out that the only explicit linkage marking here is `extern`, not `static`, but you're certainly correct that the extern applies when relocatable device code is enabled, and the `static` comment refers to when it is disabled. The comment still seems a bit stale, however. because the "we make this clear by marking it static" isn't true any longer.

> Regarding the inline thing: a pure inline didn't work for some weird reason I don't remember. Essentially we had in the field errors we tracked down to this @ibaned may remember more about it. Most of the problems were anyway specific to NVCC and not Clang ...

It would be interesting to know whether this comment refers to when relocatable device code is enabled or not. I'm guessing not (as that's the only case where there are multiple possible addresses from which to choose). Functions with inline linkage generally can't refer to static globals because at link time you get only one non-inlined copy of the inline function (and so it will refer to one of static globals, but in general one in the wrong translation unit). My thought when looking at this code was: Ah, this could be a function if you declared its linkage as:

```
#ifdef KOKKOS_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE
inline
#else 
static
#endif
```

but maybe I'm misunderstanding the problem.

**Comment by crtrott** [comment]
Yeah that could be true.

**Comment by crtrott** [comment]
We should fix the comment.

**Comment by cwpearson** [comment]
#5817 removed the CUDA and HIP lock arrays
